### PR TITLE
Fixes #799: Exposes collector plugin defaults to the framework

### LINKED ISSUE #799: When a plugin exposes config with defaults those defaults need to be available to framework
For instance, the call to GetMetricTypes is called with config from the pluginConfig which does not include the defaults that were exposed by the plugin.

### PR DESCRIPTION
**Pull request by tiffanyfj**
Fixes #799 

Summary of changes:
- Switched ctypes.ConfigValueX to pass by value instead of reference
- Updated mock1 so the required password doesn't fail all tests since I moved the rules to the root
- Added defaults to GetMetricTypes call and the plugin config

Testing done:
- Ran tests
- Added testing
- Removed check from Apache plugin and it loaded correctly
- Loaded plugins with print statements
     - For mock1 the config passed into GetMetricTypes is cfg : map[name:{bob}]
     - For CollectMetrics name is only being set in the config policy and for "test" it gives "The mock collected data! config data: name={bob} password={testval} test=true"

@intelsdi-x/snap-maintainers

### CODE REVIEW

File: control/plugin_manager.go [hunk 1]
```diff
@@ -355,10 +355,45 @@ func (p *pluginManager) LoadPlugin(details *pluginDetails, emitter gomit.Emitter
 	lPlugin.ConfigPolicy = cp
 
 	if resp.Type == plugin.CollectorPluginType {
+		cfgNode := p.pluginConfig.getPluginConfigDataNode(core.PluginType(resp.Type), resp.Meta.Name, resp.Meta.Version)
+
+		if lPlugin.ConfigPolicy != nil {
+			// Get config policy from root/plugin defaults
+			ncd := lPlugin.ConfigPolicy.Get([]string{""})
```
Comment (IRCody) [inline control/plugin_manager.go]: What does ncd stand for?

File: control/control_test.go [hunk 2]
```diff
@@ -838,6 +838,32 @@ func TestMetricConfig(t *testing.T) {
...
+			errs := c.validateMetricTypeSubscription(m1, cd)
+			So(errs, ShouldBeNil)
+		})
+		c.Config.Plugins.All.AddItem("name", ctypes.ConfigValueStr{Value: "jane"})
+		Convey("So mock should have name: bob config from defaults", func() {
+			So(c.Config.Plugins.pluginCache["0:mock:1"].Table()["name"], ShouldResemble, ctypes.ConfigValueStr{Value: "bob"})
+		})
+
+		c.Stop()
+		time.Sleep(100 * time.Millisecond)
```
Comment (IRCody) [inline control/control_test.go]: Why do we need to sleep ?

Comment (tiffanyfj) [inline control/control_test.go]: I copied it from above. So...not sure?

File: control/plugin_manager.go (hunk 1 above)
Comment (tiffanyfj) [inline control/plugin_manager.go]: Also no clue...got it from control.go

Comment (tiffanyfj) [inline control/plugin_manager.go]: Node config data? Throwing things out there... @geauxvirtual or @danielscottt would know according to git blame.

Comment (IRCody) [inline control/plugin_manager.go]: That's probably it.

**Comment by IRCody** [comment]
LGTM

File: control/plugin_manager.go (hunk 1 above)
Comment (lynxbat) [inline control/plugin_manager.go]: Yup, this was me.

File: control/plugin_manager.go [hunk 3]
```diff
@@ -355,10 +355,45 @@ func (p *pluginManager) LoadPlugin(details *pluginDetails, emitter gomit.Emitter
...
+					fields[idx] = e.Error()
+				}
+				fields["name"] = resp.Meta.Name
+				fields["version"] = resp.Meta.Version
+				se.SetFields(fields)
+				return nil, se
+			}
+
+			// Update config policy with defaults
+			cp, err = c.GetConfigPolicy()
```
Comment (jcooklin) [inline control/plugin_manager.go]: Is this block needed given https://github.com/tiffanyfj/snap/blob/00a75a229df3f72ef5a4f5c41865a55699481945/control/plugin_manager.go#L343-L355

File: control/control_test.go (hunk 2 above)
Comment (danielscottt) [inline control/control_test.go]: Can we find a way to test this without time?

File: control/plugin_manager.go (hunk 3 above)
Comment (tiffanyfj) [inline control/plugin_manager.go]: Hmm so it mentions collector in there...but I thought it is used for all types.

**Comment by tiffanyfj** [comment]
Added part of @jcooklin's commit for a future grpc PR here.

File: control/control_test.go (hunk 2 above)
Comment (tiffanyfj) [inline control/control_test.go]: Have a suggestion besides just removing it from the code and seeing how Travis etc like it?

Comment (tiffanyfj) [inline control/control_test.go]: Seems they were added for Travis issues and I just copied it from another line...

https://github.com/intelsdi-x/snap/commit/da78c23ebea8a5b183e8086e8f8e8171268dfb16
@geauxvirtual

**Comment by geauxvirtual** [comment]
LGTM.

**Pull request merged**